lib/datadog/tracing.rb

```diff
@@ -0,0 +1,198 @@
+      # @public_api
+      def continue_trace!(digest, &block)
+        tracer.continue_trace!(digest, &block)
+      end
+
+      # The currently active {Datadog::Tracer} instance.
+      #
+      # The instance returned can change throughout the lifetime of the application.
+      # This means it is not advisable to cache it.
+      #
+      # The trace can be configured through {.configure},
+      # through {Datadog::Configuration::Settings::DSL::Tracer} options.
+      #
+      # TODO: This next paragraph can be better written.
+      #
+      # Most of the functionality available through the {.tracer} instance is
+      # also available in public methods in the {Datadog::Tracing} module.
+      # It is preferable to use the public methods in the {Datadog::Tracing} when possible
+      # as {Datadog::Tracing} strongly defines the tracing public API, and thus
+      # we strive to no introduce breaking changes to {Datadog::Tracing} methods.
+      #
+      # @return [Datadog::Tracer] the active tracer
+      # @!attribute [r] tracer
+      # @public_api
+      def tracer
+        components.tracer
+      end
+
+      # The tracer's internal logger instance.
+      # All tracing log output is handled by this object.
+      #
+      # The logger can be configured through {.configure},
+      # through {Datadog::Configuration::Settings::DSL::Logger} options.
+      #
+      # @!attribute [r] logger
+      # @public_api
+      def logger
+        Datadog.logger
+      end
+
+      # Current tracer configuration.
+      #
+      # To modify the configuration, use {.configure}.
+      #
+      # @return [Datadog::Configuration::Settings]
+      # @!attribute [r] configuration
+      # @public_api
+      def configuration
+        Datadog.configuration
+      end
+
+      # Apply configuration changes to `ddtrace`. An example of a {.configure} call:
+      # ```
+      # Datadog.configure do |c|
+      #   c.sampling.default_rate = 1.0
+      #   c.use :aws
+      #   c.use :rails
+      #   c.use :sidekiq
+      #   # c.diagnostics.debug = true # Enables debug output
+      # end
+      # ```
+      #
+      # Because many configuration changes require restarting internal components,
+      # invoking {.configure} is the only safe way to change `ddtrace` configuration.
+      #
+      # Successive calls to {.configure} maintain the previous configuration values:
+      # configuration is additive between {.configure} calls.
+      #
+      # The yielded configuration `c` comes pre-populated from environment variables, if
+      # any are applicable.
+      #
+      # See {Datadog::Configuration::Settings} for all available options, defaults, and
+      # available environment variables for configuration.
+      #
+      # @yieldparam [Datadog::Configuration::Settings] c the mutable configuration object
+      # @return [void]
+      # @public_api
+      def configure(&block)
+        Datadog.configure(&block)
+      end
```

Comment:
This... is really weird.

We're making `Datadog::Tracing.configure` a shortcut for `Datadog.configure`, but that means that you can use `Datadog::Tracing.configure` to setup the profiler, or any other component, which seems wrong.

Should we only allow tracing-related configuration to be used in this way?

Reply:
So, David and I don't think we can separate profiling from tracing configuration and components in time for 1.0.

Today, both projects live together in `ddtrace`, and are configured together, despite profiling being a distinct product.
But it's not simple to create a separate `Datadog::Profiling.configure` that works separate from `Datadog::Tracing.configure`.

The compromised we settled on was to namespace everything now, then refactor to break out the separate profiling and tracing products later.

Reply:
We can keep a global `Datadog.configure` for the time being.

Reply:
I think we could have both a `Datadog::Tracing.configure` and a `Datadog::Profiling.configure` both alias to a common core (`Datadog.configure`) and have our cake and eat it too. Then privatize `Datadog.configure` and have users use the `Tracing`/`Profiling` interfaces versions instead.

Reply:
My concern here is backwards-compatibility.

If `Datadog::Tracing.configure` aliases `Datadog.configure`, then effectively we're committing for this to never change during 1.x, and thus to keep on supporting the configuration of all components through every `configure` method.

That said, I have a question and a suggestion:

1. Is the plan to hide `Datadog.configure` for 1.x right away, or only later?
2. If we plan on keeping `Datadog.configure` for 1.x (even if we deprecate it at some point), I would suggest an alternative approach: 

    We could avoid introducing `Datadog::Tracing.configure` in 1.0.0, and then work on introducing a proper product-scoped configuration during the 1.x cycle. Aka seeing as this is a new interface, we don't have to have it from 1.0.0, and can take a bit more time to iron out the details.

Reply:
My suggestion would be to immediately hide `Datadog.configure` (make it publicly inaccessible) in 1.0, and offer `Datadog::Tracing.configure` and `Datadog::Profiling.configure` as alternatives.

Using options for tracing in profiling, or vice versa would not be supported. In each of these configuration routes, filter and raise errors for inappropriate options. I imagine this could be done by wrapping the block provided to `Datadog::Profiling#configure` with a "validation" lambda, then forwarding that to `Datadog.configure`. (Same for tracing.)

Reply:
Works for me 👍

Reply:
A side note, this PR is only for the Public Tracing API, so profiling changes are out of scope. We can work towards the final goal of this discussion in separate PRs (one for profiling changes, one for global changes (`Datadog.configure`)).

Reply:
Suggestion: Should we have `Datadog::Tracing.configure` be a "raise TODO" until we revisit it in the other PRs?

Reply:
I'd prefer we keep it one way or another in some kind of functional state, between PRs, even if it isn't optimal.

Reply:
I'm concerned that if we merge this with the alias being as-is, it won't get revised in time for 1.0.0.

Hence my above suggestion of leaving it as a TODO, so we won't forget.

Reply:
Update: My comment is no longer relevant, now that the validation was added 😅